### PR TITLE
`Programming exercises`: Add blackbox tests as another Java project type

### PR DESCRIPTION
**Pull request by chrisknedl**
# WARNING: DELETE EXERCISES CREATED WITH THIS PR AFTER TESTING! OTHERWISE THE TEST SERVER IS BROKEN FOR OTHER PRs

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [ ] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

In some cases, it is preferable to leave it up to the student how their program is structured. In such cases, methods and classes of the solution are not known beforehand. In order to test such solutions, blackbox tests are required. This PR adds that functionality for Java programming exercises.

### Description

The testing framework that is used is called `DejaGnu` (https://www.gnu.org/software/dejagnu/). A `DejaGnu` button is added to the project types for Java exercises in the client. When selected, the exemplary dependency, testwise coverage and sequential test runs are disabled.   

The template files for this exercise are stored in `templates/java/maven_blackbox`. As implied by the folder name, the DejaGnu image is based on Maven. The DejaGnu template exercise adopts the sorting exercise from the other two project types and turns it into a user input based exercise.

The pipeline.groovy file that defines the Jenkins pipeline steps is located in the `templates/java/test/blackbox.projectTemplate` directory. This folder also includes all the files required for testing. The actual docker image is defined in the pipeline.groovy file (ghcr.io/uni-passau-artemis/artemis-dejagnu:20).

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Course
- TS2 (Jenkins Setup)

1. Log in to Artemis
2. Navigate to Course Management
3. Create a Java programming exercise with online editor enabled, and choose the new `DejaGnu` project type.
4. Make sure the BASE build fails, and the SOLUTION build works
5. Look inside the test-repository via the online editor, it should contain the files that are shown in the two screenshots at the bottom.

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| programming-exercise-update.component.ts | 80% | ✅              |

### Screenshots

UI change:
Bildschirmfoto 2023-06-26 um 10 23 53

Test Repo folder structure:
Bildschirmfoto 2023-06-26 um 09 38 23
Bildschirmfoto 2023-06-26 um 09 38 28
`bb` will be replaced by your package name.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseRepositoryService.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseRepositoryService.java:320]:
```suggestion
            Path dejagnuLibFolderPath = repoLocalPath.resolve("testsuite").resolve("lib");
```
That changed with Dejagnu 1.6.3, I think?

Comment (chrisknedl) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseRepositoryService.java]: True, should be part of the fixed I just pushed. Thanks for noticing.

**Review by DominikRemo — DISMISSED** [review]
Tested on ts2. Creating the new programming exercise worked well and participation as a Student also worked as expected. Nice addition 🚀

**Review by laadvo — COMMENTED** [review]
I followed the testing steps on ts2, there seems to be a problem with the solution building. I'm not sure if this is a problem with the test server or not since I cannot test locally.
image
image

**Comment by chrisknedl** [comment]
Thanks for testing. Was the secret.exp file present in the test repository?

File: src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy
Comment (b-fein) [inline src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy:160]: In case the package name contains multiple dots the toolname detection might break which might have caused the issue @laadvo mentioned.

Probable fix: use `sed` to remove the `.tests` at the end
```suggestion
        script: """find testsuite -name "*.tests" -type d -printf "%f" | sed 's#.tests$##'""",
```

File: src/main/resources/templates/java/maven_blackbox/exercise/pom.xml
Comment (b-fein) [inline src/main/resources/templates/java/maven_blackbox/exercise/pom.xml:27]:
```suggestion
                    <release>20</release>
```
Recommended alternative since Java 9.

File: src/main/resources/templates/java/maven_blackbox/exercise/Tests.txt
Comment (b-fein) [inline src/main/resources/templates/java/maven_blackbox/exercise/Tests.txt:1]: Note to other reviewers: Intentionally empty as part of the template. The student should fill it with content.

File: src/main/resources/templates/java/maven_blackbox/exercise/pom.xml
Comment (b-fein) [inline src/main/resources/templates/java/maven_blackbox/exercise/pom.xml:24]:
```suggestion
                <version>3.11.0</version>
```
Dependency can be updated.

File: src/main/resources/templates/java/maven_blackbox/readme
Comment (b-fein) [inline src/main/resources/templates/java/maven_blackbox/readme:125]:
```suggestion
### Part 5: Optional Challenges
```
Section numbering

File: src/main/resources/templates/java/maven_blackbox/solution/pom.xml
Comment (b-fein) [inline src/main/resources/templates/java/maven_blackbox/solution/pom.xml:27]:
```suggestion
                <version>3.11.0</version>
                <configuration>
                    <release>20</release>
```

File: src/main/resources/templates/java/maven_blackbox/solution/src/${packageNameFolder}/Client.java
Comment (b-fein) [inline src/main/resources/templates/java/maven_blackbox/solution/src/${packageNameFolder}/Client.java:14]:
```suggestion
```
Not needed. Student can figure that out.

File: src/main/resources/templates/java/maven_blackbox/solution/src/${packageNameFolder}/Client.java
Comment (b-fein) [inline src/main/resources/templates/java/maven_blackbox/solution/src/${packageNameFolder}/Client.java:44]:
```suggestion
```

File: src/main/resources/templates/java/test/blackbox/projectTemplate/pom.xml
Comment (b-fein) [inline src/main/resources/templates/java/test/blackbox/projectTemplate/pom.xml:26]:
```suggestion
                    <release>20</release>
```

File: src/main/resources/templates/java/test/blackbox/projectTemplate/readme.md
Comment (b-fein) [inline src/main/resources/templates/java/test/blackbox/projectTemplate/readme.md:5]:
```suggestion
#### Project structure
The `testsuite` directory contains the test definitions used by DejaGnu.
`config/default.exp` contains the base setup shared by all tests.
The actual tests are split up into `public.exp`, `advanced.exp`, and `secret.exp`.

In the `testfiles/` additional support files can be provided that can be loaded by the program under test.
The `secret/` directory can contain files that should only be readable during execution of the `secret` tests, but not during the other two.

To make the `secret` tests actually secret (i.e. results not visible to the student), you need to configure the test visibility in the grading configuration of the exercise.
```
Adapted to DejaGnu.

File: src/main/resources/templates/java/test/blackbox/projectTemplate/testsuite/${packageNameFolder}.tests/advanced.exp
Comment (b-fein) [inline src/main/resources/templates/java/test/blackbox/projectTemplate/testsuite/${packageNameFolder}.tests/advanced.exp:3]:
```suggestion
# Tests for advanced functionality
```
The advanced tests are still visible to the student, usually.

File: src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy
Comment (b-fein) [inline src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy:12]:
```suggestion
dockerImage = "#dockerImage"
dockerFlags = "#dockerArgs"
```
Should be variables set by the Jenkins build plan creator

File: src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy
Comment (b-fein) [inline src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy:15]:
```suggestion
isStaticCodeAnalysisEnabled = #isStaticCodeAnalysisEnabled
```
Unused, and instead to be in line with the other templates

File: src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy
Comment (b-fein) [inline src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy:241]:
```suggestion
private void staticCodeAnalysis() {
    if (!isStaticCodeAnalysisEnabled) {
        return
    }

```

**Review by b-fein — APPROVED** [review]
Code review: only smaller suggestions, mostly to adapt the documentation to the new project type.

File: src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy
Comment (laadvo) [reply src/main/resources/templates/jenkins/java/blackbox/regularRuns/pipeline.groovy:160]: I think this might be the issue. I used a package name with multiple dots when testing

**Review by b-fein — CHANGES_REQUESTED** [review]
Warning message when package name is invalid is misleading. The regular Java one is shown even though no `.` are allowed:
image

Otherwise code looks good. Will test on TS2 properly when its connected GitLab works again.

**Comment by chrisknedl** [comment]
@b-fein Thanks for the review. I'm trying to fix the message right now.

**Review by b-fein — DISMISSED** [review]
Tested on TS2: Created a new exercise, submitted as student, recreated build plans. Everything worked without issues.

The error message noted in the previous review has been fixed.

Code: looks good as well.

Thank you!

**Comment by chrisknedl** [comment]
@b-fein Thanks a lot for both testing and the review!

**Review by b-fein — DISMISSED** [review]
Reapprove after minimal merge-conflict only in `application.yml`.

**Review by DominikRemo — APPROVED** [review]
Reapproved. Tested on ts2

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (laadvo) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:87]: Here is another occurence of it. The regex is for the blackbox package name, so I'm not sure if blackbox project type is correct here in the comment.

**Review by laadvo — COMMENTED** [review]
Code looks good, a really cool addition!
I only had one question regarding the "project type"

File: src/main/webapp/i18n/en/exercise.json
Comment (chrisknedl) [reply src/main/webapp/i18n/en/exercise.json:223]: What I meant with that was: For all other project types, dots are usually allowed as separators in the package name, but in the specific case of the blackbox type, they aren't. I guess `In addition, no dots are allowed in the package name of blackbox projects!` would make it more clear?

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (chrisknedl) [reply src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:87]: Also here: `No dots allowed in package names when the blackbox project type is selected`?

File: src/main/webapp/i18n/en/exercise.json
Comment (b-fein) [reply src/main/webapp/i18n/en/exercise.json:223]: I’d prefer the ‘for the blackbox project type’ as-is. The UI already calls the choice Gradle/Maven/Dejagnu the ‘Project Type’.

Follow-up question: Should that then refer to ‘DejaGnu’ tests in the warning message rather than blackbox-tests so that the user can mentally connect those two things? Otherwise they might now know that DejaGnu = Blackbox.

Comment (laadvo) [reply src/main/webapp/i18n/en/exercise.json:223]: `In addition, no dots are allowed in the package name of blackbox projects!` would make it very clear :+1:

Comment (laadvo) [reply src/main/webapp/i18n/en/exercise.json:223]: But it's optinal, I now understand it. I agree with @b-fein, writing DejaGnu would make sense since the project type is called that

Comment (b-fein) [reply src/main/webapp/i18n/en/exercise.json:223]:
```suggestion
                        "JAVA_BLACKBOX": "The package name must be a valid Java identifier. In addition, no dots are allowed in the package name of DejaGnu projects"
```

File: src/main/webapp/i18n/de/exercise.json
Comment (b-fein) [inline src/main/webapp/i18n/de/exercise.json:223]:
```suggestion
                        "JAVA_BLACKBOX": "Der Package-Name muss ein gültiger Java-Bezeicher sein. Zusätzlich sind Trennpunkte im Paketnamen für den DejaGnu Aufgabentyp nicht erlaubt."
```

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (chrisknedl) [reply src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:87]: But the variable name `packageNamePatternForJavaKotlin` already encodes that piece of information, so it should be fine here as is.

Comment (laadvo) [reply src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:87]: Here is really optional, since it's just a comment. Feel free to leave it as it is

**Comment by chrisknedl** [comment]
Thanks everyone!

**Review by laadvo — APPROVED** [review]
Thanks for answering my question! Code looks good to me

**Review by b-fein — DISMISSED** [review]
Code and functionality re-approve after only a small change to the translation strings.

**Review by DominikRemo — APPROVED** [review]

**Review by b-fein — APPROVED** [review]
Only small merge conflict in the `JenkinsProgrammingLanguageFeatureService` due to the auxiliary repositories being disabled with #7122.

Reapprove.

**Review by laadvo — APPROVED** [review]
Reapprove code after merge conflict

**Comment by MarkusPaulsen** [comment]
Please try to get at least 4 approvals before setting the ready to merge label.

**Review by DominikRemo — DISMISSED** [review]

**Comment by Strohgelaender** [comment]
I don't see in the changes where the build plan for this new type gets created. Since this is also a maven project, I assume that the normal maven build plan just gets used here. However, `isMavenProject()` returns false for this type (which in itself is confusing for a type called `maven_blackbox`). Does this just rely on a `if (isGradle()) {} else {} ` code in the repository? 

At a minimum, I would add a comment at the respective place explaining that this brach must cover 3 project types.

In general, I'd like to have `isMavenProject()` return true for this new type, since it really is a maven project. For places where not the same logic can be applied, an explicit check like `if (type == MAVEN_BLACKBOX) { } else if (isMavenProject(type)) { }` would be better.

File: src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/build_plan/JenkinsPipelineScriptCreator.java
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/build_plan/JenkinsPipelineScriptCreator.java:125]: @Strohgelaender This choice selects the relevant subdirectory from which the build script template should be loaded.

Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/build_plan/JenkinsPipelineScriptCreator.java:125]: ok, thanks. So it is an explicit check (I thought this was selecting the template folder when looking through the changes).

**Comment by b-fein** [comment]
> Does this type need the special command-pattern-exercise template?

Do you mean the different `Client.java` and its dependencies in the `input` package? If so, yes they are needed. The idea of this template is to be able to test programs as blackboxes via their CLI rather than unit-tests. This requires a different `Client` implementation, unless the `Client` is changed in a future separate PR to accept a command line parameter that switches from non-interactive to interactive mode.

**Review by Strohgelaender — APPROVED** [review]
thanks for answering the questions, lgtm now.

**Review by laadvo — APPROVED** [review]
reapprove code

**Review by b-fein — APPROVED** [review]
Re-tested exercise creation and project import after.

**Review by DominikRemo — APPROVED** [review]

**Review by MarkusPaulsen — APPROVED** [review]
Code looks good.

**Pull request merged**
